From e58355ab59c95b3ce708d3fdb417e3893170cfde Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Tue, 20 Aug 2019 14:34:57 +0200
Subject: [PATCH 06/14] net: stmmac: sun8i: Fix reboot issue with ethernet on
 Orange Pi 3

We disable both phy regulators before reboot, so that they are in a
known state on the next boot. This should be done in u-boot, but
it doesn't support AXP806 yet, thus this workaround.

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 .../net/ethernet/stmicro/stmmac/dwmac-sun8i.c | 22 +++++++++++++++++++
 1 file changed, 22 insertions(+)

--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-sun8i.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-sun8i.c
@@ -17,6 +17,7 @@
 #include <linux/phy.h>
 #include <linux/platform_device.h>
 #include <linux/regulator/consumer.h>
+#include <linux/reboot.h>
 #include <linux/regmap.h>
 #include <linux/stmmac.h>
 
@@ -80,6 +81,7 @@ struct sunxi_priv_data {
 	struct regmap_field *regmap_field;
 	bool internal_phy_powered;
 	void *mux_handle;
+	struct notifier_block reboot_nb;
 };
 
 /* EMAC clock register @ 0x30 in the "system control" address range */
@@ -1130,6 +1132,19 @@ out_put_node:
 	return regmap;
 }
 
+
+static int sun8i_dwmac_reboot_notifier(struct notifier_block *nb,
+				       unsigned long action, void *data)
+{
+	struct sunxi_priv_data *gmac = container_of(nb, struct sunxi_priv_data,
+						    reboot_nb);
+
+	regulator_disable(gmac->regulator_phy);
+	regulator_disable(gmac->regulator_phy_io);
+
+	return NOTIFY_DONE;
+}
+
 static int sun8i_dwmac_probe(struct platform_device *pdev)
 {
 	struct plat_stmmacenet_data *plat_dat;
@@ -1184,6 +1199,13 @@ static int sun8i_dwmac_probe(struct plat
 		return ret;
 	}
 
+	gmac->reboot_nb.notifier_call = sun8i_dwmac_reboot_notifier;
+	ret = devm_register_reboot_notifier(dev, &gmac->reboot_nb);
+	if (ret) {
+		dev_err(dev, "Failed to register reboot notifier (%d)\n", ret);
+		return ret;
+	}
+
 	/* The "GMAC clock control" register might be located in the
 	 * CCU address range (on the R40), or the system control address
 	 * range (on most other sun8i and later SoCs).
